pypiserver/backend.py

```diff
@@ -167,6 +167,15 @@ def __init__(
 
         self.cache_manager = cache_manager or CacheManager()  # type: ignore
 
+    def add_package(self, filename: str, stream: t.BinaryIO) -> None:
+        write_file(stream, self.roots[0].joinpath(filename))
+        self.cache_manager.invalidate_root_cache(self.roots[0])
+
+    def remove_package(self, pkg: PkgFile) -> None:
+        if pkg.fn is not None:
```

Comment:
is this ever called with `None`?

Reply:
Honestly I copied these straight out of the `SimpleFileBackend` and added the cache manager calls without thinking too much about it. I probably should update to calling `super().add_package(...)` and then update the ones there to address these concerns

Reply:
I don't know if this ever gets called with `pkg.fn` being none. It's allowed to be none according to the type definition of `PkgFile`, but it does seem like we'd need something like this. I'll dig in a bit and see if this check is really necessary, or if we can maybe change the type of the `PkgFile` class

Reply:
Haha, then it must've been my doing :roll_eyes: . I guess I did the None-check to make the typechecker happer.

I'm currently working on simplifying `PkgFile`, so will have a look at it as well